autoload/ale/statusline.vim

```diff
@@ -1,29 +1,54 @@
 " Author: KabbAmine <amine.kabb@gmail.com>
 " Description: Statusline related function(s)
 
-function! ale#statusline#Status() abort
-    " Returns a formatted string that can be integrated in the
-    " statusline
-
-    let l:buffer = bufnr('%')
-    let l:loclist = g:ale_buffer_loclist_map
-
-    if !has_key(l:loclist, l:buffer)
-        return ''
-    endif
-
+" Update the buffer error/warning count with data from loclist.
+function! ale#statusline#Update(buffer, loclist) abort
     let l:errors = 0
     let l:warnings = 0
-    for l:entry in l:loclist[l:buffer]
+
+    for l:entry in a:loclist
         if l:entry.type ==# 'E'
             let l:errors += 1
         else
             let l:warnings += 1
         endif
     endfor
 
-    let l:errors = l:errors ? printf(g:ale_statusline_format[0], l:errors) : ''
-    let l:warnings = l:warnings ? printf(g:ale_statusline_format[1], l:warnings) : ''
+    let g:ale_buffer_count_map[a:buffer] = [l:errors, l:warnings]
+endfunction
+
+" Returns a tuple of errors and warnings (or false if no data exists)
+" for use in third-party integrations.
+function! ale#statusline#Count(buffer) abort
+    if !has_key(g:ale_buffer_count_map, a:buffer)
+        if has_key(g:ale_buffer_loclist_map, a:buffer)
+            call ale#statusline#Update(a:buffer, g:ale_buffer_loclist_map[a:buffer])
+            return ale#statusline#Count(a:buffer)
+        else
+            return 0
+        endif
```

Comment:
You could do this instead.

```vim
if !has_key(g:ale_buffer_count_map, a:buffer)
    call ale#statusline#Update(a:buffer, get(g:ale_buffer_loclist_map, a:buffer, []))
endif

return g:ale_buffer_count_map[a:buffer]
```

The data will be replaced when there are actually errors or warnings to display, and you will get `[0, 0]` when there is no data, so there won't be any difference between the initial load and when ALE doesn't find problems later on.